jax/interpreters/masking.py

```diff
@@ -104,30 +106,98 @@ class ShapeExpr(tuple):  # type ShapeExpr = [Poly]
   def __str__(self):
     return 'ShapeExpr({})'.format(', '.join(map(str, self)))
   def __getitem__(self, idx):
-    if type(idx) is int:
-      return super(ShapeExpr, self).__getitem__(idx)
-    else:
-      return ShapeExpr(super(ShapeExpr, self).__getitem__(idx))
+    poly = super(ShapeExpr, self).__getitem__(idx)
+    return poly if type(idx) is int else ShapeExpr(poly)
+
+
+def canonicalize_poly(p):
+  if onp.isscalar(p) and onp.issubdtype(onp.int, onp.array(p).dtype):
+    return constant_poly(p)
+
+  assert isinstance(p, Poly)
+  return p
+
+def canonicalize_polymorphic_shape(shape):
+  return ShapeExpr(map(canonicalize_poly, shape))
+
+def poly_without_zeros(d):
+  d = {mon: count for mon, count in d.items() if count != 0}
+
+  return constant_poly(0) if len(d) == 0 else Poly(d)
 
 class Poly(Counter):  # type Poly = Map Mon Int -- monomials to coeffs
-  def __mul__(p1, p2):
-    new_poly = Poly()
-    for (mon1, coeff1), (mon2, coeff2) in it.product(p1.items(), p2.items()):
+  def __add__(self, other):
+    d = self.copy()
+
+    for mon, count in canonicalize_poly(other).items():
+      d[mon] = d.get(mon, 0) + count
+
+    return poly_without_zeros(d)
+
+  def __sub__(self, other):
+    return self + -other
+
+  def __neg__(self):
+    return Poly({mon: -count for mon, count in self.items()})
+
+  def __mul__(self, other):
+    new_poly = dict()
+    for (mon1, coeff1), (mon2, coeff2) \
+            in it.product(self.items(), canonicalize_poly(other).items()):
       mon = Mon(mon1 + mon2)                        # add monomials' id degrees
       coeff = coeff1 * coeff2                       # multiply integer coeffs
       new_poly[mon] = new_poly.get(mon, 0) + coeff  # accumulate coeffs
-    return new_poly
 
-  def __add__(p1, p2):
-    return Poly(Counter.__add__(p1, p2))
+    return poly_without_zeros(new_poly)
+
+  def __rmul__(self, other):
+    return self * other
+
+  def __radd__(self, other):
+    return self + other
+
+  def __rsub__(self, other):
+    return self + -other
+
+  def __floordiv__(self, divisor):
+    q, _ = divmod(self, divisor)
+    return q
+
+  def __mod__(self, divisor):
+    _, r = divmod(self, divisor)
+    return r
+
+  def __divmod__(self, divisor):
+    if self.is_constant:
+      q, r = divmod(int(self), divisor)
+
+      return constant_poly(q), r
+
+    def divided(count):
+      q, r = divmod(count, divisor)
+      if r != 0:
+        raise ValueError('shapecheck currently only supports strides '
```

Comment:
I think we can support division when there's also an added constant, i.e. `c * poly + k` can be safely integer-divided by `c`. If that's right, can you add a note or a todo here about being able to support that case? If not, teach me why not :)

Reply:
Great point! Will add this.

Reply:
This is now implemented.